awscli/autoprompt/core.py

```diff
@@ -10,9 +10,20 @@
 # distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
+import sys
```

Comment:
Alright so I have been doing some thinking on how to organize these changes. I think most of the logic needed is here and most of my feedback is about rearranging the interfaces. Specifically, it seems like this change was optimized toward reducing the change in the top-level `awscli.clidriver.main()` function. I think the problem with taking this approach is that as a result we have to push a lot of the logic into the `AutopromptDriver` and ends up seeming like it does too much, is not too reusable (e.g. similar to where we are at with the CLIDriver), and we also have to introduce new non-integer return codes to get it all working.

I think instead what we should do is create a new top-level class the wires together the logic of these two major classes (`CLIDriver` and `AutopromptDriver`) to make both of these classes less bloated and more reusable. Let's call this class for now `AWSCLIEntryPoint` and it has this sample structure:
```python
class AWSCLIEntryPoint:  # Maybe we can come with a better class name
    def main(self, args):
        try:
            rc = self._do_main(args)
        except Exception as e:
            # log exception
            # Set RC based on whatever handlers we want here
        HISTORY_RECORDER.record('CLI_RC', rc, 'CLI')
        return rc

    def _do_main(self, args):
        driver = create_clidriver()
        autoprompt_driver = AutoPromptDriver(driver)
        # This will return the mode and throw any validation errors if needed
        auto_prompt_mode = autoprompt_driver.resolve_mode(args)
        # We can also probably isolate each if branch to separate methods
        # and delegate to them using a gettattr()
        if auto_prompt_mode == 'on':
            args = autoprompt_driver.prompt_for_args(args)
            rc = driver.main(args)
        elif auto_prompt_mode == 'on-partial':
            autoprompt_driver.inject_requires_prompting_error_handler(driver)
            try:
                rc = driver.main(args)
            except autoprompt_driver.RequiresPromptingException:
                args = autoprompt_driver.prompt_for_args(args)
                driver = create_clidriver()
                rc = driver.main(args)
        else:
            rc = driver.main(args)
        return rc
```
and then make `main()` simply:
```python
def main():
    return AWSCLIEntryPoint().main(sys.argv[1:])
```
I sort of like this approach because:
1. It makes it easier to follow what the flow of a CLI command is from a single class. As it is right now, it sort of tricky to follow how the behavior works because this logic is all encompassed in the `AutoPromptDriver.prompt_for_args()` and that method is a little misleading itself in that the CLIDriver may be ran on your behalf (and succeed) or the user may never be auto-prompted for.
2. It simplifies the role of the `AutoPromptDriver` and its interfaces to help narrow down its role: as a bridge between CLI parameters/configuration and the `AutoPrompt` class to control when to use it. In terms of the interfaces that I had in the code sample, I just added two more methods: 1) `resolve_mode` which just does the validation of arguments and returns what mode to use. 2) `inject_requires_prompting_error_handler` which just should prepend a handler that raises a `RequiresPromptingException` in the chain handler if there is a parameter validation error. Then the `prompt_for_args` method should be as simple as delegating to the `AutoPrompter`.
3. The class starts laying out the framework for making the `CLIDriver` more pluggable and reusable. For example, on top of these changes I would like to see in the future we introduce a factory/builder for setting up a session based on configuration/global arguments that can be plumbed into the `CLIDriver`.  It also allows for better testing in that we can also add a `cli_driver_factory` and `auto_prompt_factory` to the tantalizer so we do not necessarily have to rely on patching. 
4. We keep `rc` reserved to integers/values that users will receive in their shell instead of introducing other potential type of rc's that we would need to account for and trip us up.

That was a bit long, but I don't think it is too big of a change of direction from where we are at. Let me know what you think.

Reply:
👍